File: src/main/java/im/threads/business/database/DatabaseHolder.java
package im.threads.business.database;

import androidx.annotation.NonNull;

import java.util.List;

import im.threads.business.config.BaseConfig;
import im.threads.business.models.ChatItem;
import im.threads.business.models.FileDescription;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

public final class DatabaseHolder {

    private static DatabaseHolder instance;
    private final ThreadsDbHelper mMyOpenHelper;

    private DatabaseHolder() {
        mMyOpenHelper = new ThreadsDbHelper(BaseConfig.instance.context);
    }

    @NonNull
    public static DatabaseHolder getInstance() {
        if (instance == null) {
            instance = new DatabaseHolder();
        }
        return instance;
    }

    @NonNull
    public List<ChatItem> getChatItems(int offset, int limit) {
        return mMyOpenHelper.getChatItems(offset, limit);
    }

    public Single<List<FileDescription>> getAllFileDescriptions() {
        return Single.fromCallable(mMyOpenHelper::getAllFileDescriptions)
                .subscribeOn(Schedulers.io());
    }
}

File: src/main/java/im/threads/business/database/DatabaseHolder.kt
package im.threads.business.database

import im.threads.business.config.BaseConfig
import im.threads.business.models.ChatItem
import im.threads.business.models.FileDescription
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

class DatabaseHolder private constructor() {
    private val mMyOpenHelper: ThreadsDbHelper = ThreadsDbHelper(BaseConfig.instance.context)

    fun getChatItems(offset: Int, limit: Int): List<ChatItem> {
        return mMyOpenHelper.getChatItems(offset, limit)
    }

    val allFileDescriptions: Single<List<FileDescription>>
        get() = Single.fromCallable { mMyOpenHelper.allFileDescriptions }
            .subscribeOn(Schedulers.io())

    companion object {
        var instance: DatabaseHolder? = null
            get() {
                if (field == null) {
                    field = DatabaseHolder()
                }
                return field
            }
            private set
    }
}
